graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/MongoDBPreflightCheck.java

```diff
@@ -43,48 +46,66 @@ public class MongoDBPreflightCheck implements PreflightCheck {
     private final MongoConnection mongoConnection;
 
     @Inject
-    public MongoDBPreflightCheck(
-            @Named(value = "mongodb_version_probe_attempts") int mongoVersionProbeAttempts,
-            MongoConnection mongoConnection
-
-            ) {
+    public MongoDBPreflightCheck(@Named(value = "mongodb_version_probe_attempts") int mongoVersionProbeAttempts,
+                                 MongoDbConfiguration configuration) {
         this.mongoVersionProbeAttempts = mongoVersionProbeAttempts;
-        this.mongoConnection = mongoConnection;
+        // We build our own MongoConnection instance here, so we can close it without interfering with other users
+        this.mongoConnection = new MongoConnectionImpl(configuration);
     }
 
     @Override
     public void runCheck() throws PreflightCheckException {
         try {
-            final Version mongoVersion = RetryerBuilder.<Version>newBuilder()
-                    .retryIfResult(Objects::isNull)
-                    .retryIfExceptionOfType(MongoTimeoutException.class)
-                    .retryIfRuntimeException()
-                    .withRetryListener(new RetryListener() {
-                        @Override
-                        public <V> void onRetry(Attempt<V> attempt) {
-                            if (attempt.hasResult()) {
-                                return;
-                            }
-                            if (mongoVersionProbeAttempts == 0) {
+            final ImmutablePair<Version, Boolean> mongoVersion = getDbVersionAndStatus(false);
+            final Version version =  mongoVersion.getKey();
+            MongoDBVersionCheck.assertCompatibleVersion(version);
+            LOG.info("Connected to {}MongoDB with version {}", mongoVersion, mongoVersion.getValue() ? "an empty " : "");
+        } catch (ExecutionException | RetryException e) {
+            throw new PreflightCheckException("Failed to retrieve MongoDB version.", e);
+        }
+    }
+
+    public boolean dbIsEmpty() {
+        try {
+            final Boolean isEmpty = getDbVersionAndStatus(true).getValue();
+            return isEmpty;
+        } catch (ExecutionException | RetryException e) {
+            return false;
+        }
+    }
+
+    private ImmutablePair<Version, Boolean> getDbVersionAndStatus(boolean runSilent) throws ExecutionException, RetryException {
```

Comment:
Not sure if I would not split it, right now org.graylog2.bootstrap.ServerBootstrap#runPreFlightChecks will exectue this register a fresh registration without caring for the verison, afterwards its exectued as part of the preflight check, which only asserts the version but not if its empty

Reply:
I agree, this isn't ideal. Let me refactor that.